### DISCUSSION TITLE
User Aliases

### DISCUSSION THREAD

**Discussion by OverzealousLotus** [uBlockOrigin/uBlock-issues#2701] [Ideas]
- TL;DR: User-Defined Aliases used for short hands of selector chains or operations in general.

# Why?
Sometimes, filters get very long, or some selector chains are repeated in several filters. For example, in [this issue](https://github.com/uBlockOrigin/uBlock-issues/issues/1211), this type of feature is already possible according to the comments, but its syntax is extremely unreadable for someone like me. If you are able to read it, I think we can all agree that it's lengthy and gross. Aliases wouldn't remove some of its problems, but when actually used in filters, having `classes()` as a user-defined Alias would be a quality of life change. On top of this, it would remove the need for adding more selectors referring to already existing ones, which gorhill commented about in the issue linked above. It would benefit both users and contributors alike. Another example would be me, I have several filters which use the `has-text()` and `upward()` selectors respectively. I could define something like `upward-text()` as my own Alias. 

# How?
I do not think definition in the `My Filters` tab would be good, and allocating an entirely different tab for this might be a bit too much. Maybe this can be done in `Settings`, a sub-tab of `My Filters`, or an external source in `Advanced Settings`. Preferably something simple, for both gorhill and users themselves.

# Why Not?
I do not know how complex this would be to implement, as I am not experienced at all in Javascript. It could also take a toll on uBlock Origin's weight, especially for selectors which require input, such as a `classes()` alias. This would probably require variadic functions. Another issue I mentioned above, is space. I have no idea where these aliases should be defined. If anyone takes this into consideration, I thank you for reading this lengthy idea of mine. It popped into my head around two hours ago, and I really felt the need to get it out there after reading DandelionSprout's issue. I hope to hear some feedback soon!
